magnet_coils.py

```diff
@@ -0,0 +1,317 @@
+        # defined normal
+
+        # Cross-section initially populated with normal oriented along z
+        # axis
+        z_axis = np.array([0, 0, 1])
+        # Define axis of rotation as orthogonal to both z axis and surface
+        # normal
+        rot_axis = np.cross(z_axis, norm)
+        # Compute angle by which to rotate cross-section to orient along
+        # defined surface normal
+        dot_prod = np.inner(z_axis, norm)
+        rot_ang_norm = np.arccos(dot_prod/norm_mag)
+
+        # Covert angle from radians to degrees
+        rot_ang_norm = np.rad2deg(rot_ang_norm)
+
+        # Orient cross-section along defined normal
+        cubit.cmd(
+            f'rotate Surface {surf_id} angle {rot_ang_norm} about '
+            f'origin 0 0 0 direction {" ".join(str(i) for i in rot_axis)} '
+            f'include_merged'
+        )
+
+        # Conditionally orients rectangular cross-section
+        if shape == 'rectangle':
+            orient_rectangle(path[0], surf_id, norm, rot_axis, rot_ang_norm)
+            
+        # Move cross-section to initial path point
+        cubit.cmd(
+            f'move Surface {surf_id} location vertex {path[0]} '
+            f'include_merged'
+        )
+
+        # Sweep cross-section to create magnet coil
+        cubit.cmd(
+            f'sweep surface {surf_id} along curve {curve_id} '
+            f'individual'
+        )
+        # Delete extraneous curves and vertices
+        cubit.cmd(f'delete curve {curve_id}')
+        cubit.cmd('delete vertex all')
+
+        # Reinitialize path list
+        path = []
+    
+    # Delete original cross-section
+    cubit.cmd(f'delete surface {cs_id}')
+
+
+def extract_filaments(file, start, stop):
+    """Extracts filament data from magnet coil data file.
+    
+    Arguments:
+        file (str): path to magnet coil data file.
+        start (int): index for line in data file where coil data begins.
+        stop (int): index for line in data file where coil data ends (defaults
+            to None).
+
+    Returns:
+        filaments (list of list of list of float): list filament coordinates.
+            Each filament is a list of coordinates.
+    """
+    # Initialize list of filaments
+    filaments = []
+
+    # Load magnet coil data
+    file_obj = open(file, 'r')
+    data = file_obj.readlines()[start:stop]
+
+    # Initialize list of coordinates for each filament
+    coords = []
+
+    # Extract magnet coil data
+    for line in data:
+        # Parse line in magnet coil data
+        line = line.strip()
+        columns = line.split()
+
+        # Break at indicator signaling end of data
+        if columns[0] == 'end':
+            break
+
+        # Extract coordinates for vertex in path
+        x = float(columns[0])*100
+        y = float(columns[1])*100
+        z = float(columns[2])*100
+        # Extract coil current
+        s = float(columns[3])
+
+        # Coil current of zero signals end of filament
```

Comment:
This gets reset before adding the newest coordinates so that the final filament point is not appended to the list. To create a continuous path in Cubit, the initial and final points in the curve must be the same vertex. The documentation I wrote is actually incorrect, the issue is not that the coordinates must be equal, it's that the vertex ID must be equal. So, in the `create_magnets` method, I ensure the final curve vertex for each filament is the same as the initial. I will update the documentation to reflect this.